File: src/scenes/play/Invader.cpp
#include "Invader.hpp"
#include <CppInvaders.hpp>

#define TIME_DYING 0.5f

int Invader::dx = 2, Invader::dy = 0;

Invader::Invader(int i) {
    int row = 4 - i / 11;
    int col = i % 11;

    state = DOWN;
    type = row / 2;
    x = 32 + 16 * col;
    y = 64 + 16 * row;
}

Pico_Abs_Rect Invader::get_rect() const {
    switch (type) {
    case 0:
        return {x, y, 8, 8};
    case 1:
        return {x, y, 11, 8};
    default:
        return {x, y, 12, 8};
    }
}

const char* Invader::get_image() const {
    switch (type) {
    case 0:
        return IMG_INV1;
    case 1:
        return IMG_INV2;
    default:
        return IMG_INV3;
    }
}

bool Invader::is_alive() const {
    return state == UP || state == DOWN;
}

void Invader::kill() {
    state = DYING;
    timer = 0;
}

void Invader::update(float delta) {
    switch (state) {
    case DOWN:
        state = UP;
        x += dx;
        y += dy;
        break;
    case UP:
        state = DOWN;
        x += dx;
        y += dy;
        break;
    case DYING:
        timer += delta;
        if (timer >= TIME_DYING) {
            state = DEAD;
        }
        break;
    case DEAD:
        break;
    }

}

void Invader::draw() const {
    if (state == DEAD) return;

    const char* img = get_image();
    Pico_Abs_Rect r = get_rect();
    float draw_x = (float)x;
    float draw_y = (float)y;
    float w = (float)r.w;
    float h = (float)r.h;

    char key[256];
    Pico_Rel_Rect crop = { '!', {0, 0, w, h}, PICO_ANCHOR_NW, NULL };

    switch (state) {
    case UP:
        snprintf(key, sizeof(key), "/crop/invader/%d/up", type);
        crop.x = w;
        pico_layer_image_mode('=', img, img);
        pico_layer_sub_mode('=', key, img, &crop);
        break;
    case DOWN:
        snprintf(key, sizeof(key), "/crop/invader/%d/down", type);
        crop.x = 0;
        pico_layer_image_mode('=', img, img);
        pico_layer_sub_mode('=', key, img, &crop);
        break;
    case DYING:
        snprintf(key, sizeof(key), "/crop/invader/dying");
        img = IMG_EXP1;
        // Explosion frame 0
        w = 16.0f; h = 8.0f; // Assumed explosion size
        crop.w = w; crop.h = h;
        pico_layer_image_mode('=', img, img);
        pico_layer_sub_mode('=', key, img, &crop);
        break;
    default:
        return;
    }

    Pico_Rel_Rect dst = { '!', {draw_x, draw_y, w, h}, PICO_ANCHOR_N, NULL };
    pico_output_draw_layer(key, &dst);
}

File: lua/scenes/play/Invader.lua
local class = require "class"
local assets = require "assets"

local Invader = class.new()
Invader.dx = 2
Invader.dy = 0

function Invader:init(i)
    local row = 4 - math.floor(i / 11)
    local col = i % 11

    self.state = "DOWN"
    self.type = math.floor(row / 2)
    self.x = 32 + 16 * col
    self.y = 64 + 16 * row
end

function Invader:get_rect()
    if self.type == 0 then return {x=self.x, y=self.y, w=8, h=8}
    elseif self.type == 1 then return {x=self.x, y=self.y, w=11, h=8}
    else return {x=self.x, y=self.y, w=12, h=8} end
end

function Invader:get_image()
    if self.type == 0 then return assets.IMG_INV1
    elseif self.type == 1 then return assets.IMG_INV2
    else return assets.IMG_INV3 end
end

function Invader:is_alive()
    return self.state == "UP" or self.state == "DOWN"
end

function Invader:kill()
    self.state = "DYING"
    self.timer = 0
end

function Invader:update(delta)
    if self.state == "DOWN" then
        self.state = "UP"
        self.x = self.x + Invader.dx
        self.y = self.y + Invader.dy
    elseif self.state == "UP" then
        self.state = "DOWN"
        self.x = self.x + Invader.dx
        self.y = self.y + Invader.dy
    elseif self.state == "DYING" then
        self.timer = self.timer + delta
        if self.timer >= 0.5 then
            self.state = "DEAD"
        end
    end
end

function Invader:draw()
    if self.state == "DEAD" then return end

    local img = self:get_image()
    local r = self:get_rect()
    local draw_x = self.x
    local draw_y = self.y
    local w = r.w
    local h = r.h

    local key = ""
    local crop = { '!', x=0, y=0, w=w, h=h, anchor='NW' }

    if self.state == "UP" then
        key = "/crop/invader/" .. self.type .. "/up"
        crop.x = w
        pico.layer.image('=', img, img)
        pico.layer.sub('=', key, img, crop)
    elseif self.state == "DOWN" then
        key = "/crop/invader/" .. self.type .. "/down"
        crop.x = 0
        pico.layer.image('=', img, img)
        pico.layer.sub('=', key, img, crop)
    elseif self.state == "DYING" then
        key = "/crop/invader/dying"
        img = assets.IMG_EXP1
        w = 16.0; h = 8.0
        crop.w = w; crop.h = h
        pico.layer.image('=', img, img)
        pico.layer.sub('=', key, img, crop)
    else
        return
    end

    local dst = { '!', x=draw_x, y=draw_y, w=w, h=h, anchor='N' }
    pico.output.draw.layer(key, dst)
end

return Invader
